### PR TITLE
Upgrade to Cython 3+ in building

### LINKED ISSUE #56: Upgrade to Cython 3.0 and check annotations
Cython 3 has come out and there are a lot of big changes. I've run into a couple gotchas so far and we'll need to be careful. I think we should make sure to force `cython>=3` in the pyproject.toml. Issues I've run into so far:

1. DEF/constants are being deprecated. See https://github.com/cython/cython/issues/4310. However, I haven't found good replacements for some of the use cases so I'm not sure this is the first thing we tackle.
2. `cdef` functions now default to exceptions being possible (as if you had done `except -1` or similar). For many cases of the code I write this isn't what I want. On `nogil` functions this is extra bad because it adds a GIL acquire and release to check the error state. The fix is to do `noexcept nogil:` at the end of the function declaration. Checking the HTML output of `cython -a some_module.pyx` makes it pretty clear since our `nogil` functions shouldn't have any yellow (non-C code).
3. We can and should probably drop the old numpy API by defining in setup.py:

```
        define_macros = [("NPY_NO_DEPRECATED_API", "NPY_1_7_API_VERSION")]
```

### PR DESCRIPTION
**Pull request by djhoese**
This PR upgrades the build dependencies to use Cython 3+. It also adds extra annotations required by Cython 3 to avoid extra Python-related checks and stay in C-land longer and with no GIL. As far as I can tell nothing of the generated code should be different to what it was.

Note I had concerns because the modis interpolation has some low-level functions with GIL-related when memory views are returned. I made a stackoverflow question about it and a Cython maintainer responded:

https://stackoverflow.com/questions/76972950/cython-returned-memoryview-is-always-considered-uninitialized

Basically, even though it is checking the memory view for validity/initialization, it is not actually acquiring the GIL except in the failure case which should never happen for us. We could restructure the code to not create a copy of the memory views used but instead adjust the indexing appropriately. The code would be harder to read but would not have the extra memoryview checks. I doubt it would have any performance difference.

I have not run a comparison of the performance, but given the logical changes appearing in the Cython I'm not too worried. I'll have to track down my old benchmarking from the last time the Cython was touched.

 - [x] Closes #56  
 - [ ] Tests added 
 - [ ] Tests passed 
 - [ ] Passes ``git diff origin/main **/*py | flake8 --diff`` 
 - [ ] Fully documented

### CODE REVIEW

**Comment by djhoese** [comment]
So this should be the results of the same MODIS operation I did in my #38 PR:

image


I note that the memory profile looks about the same but increases more at the end than previously...but that is from other parts of Satpy, not the interpolation which is only the blue/green blocks near the beginning. I don't like the dips in CPU usage at the beginning even though execution time is about the same.

**Review by mraspaud — APPROVED** [review]
LGTM, even if I haven't checked the new features of Cython 3. Do you want to wait for an answer about the cython 3 issue you found before we merge this?

**Comment by djhoese** [comment]
Yeah let's wait a bit. I'm more OK with trollimage being released than with this being merged given the issues.

**Comment by djhoese** [comment]
FYI the bug encountered in cython seems to have been fixed, but I don't think it is released yet.

**Comment by djhoese** [comment]
Nevermind, this is another major one: https://github.com/cython/cython/issues/5688

**Comment by djhoese** [comment]
Nevermind on that one, but this one is still a thing: https://github.com/cython/cython/pull/5685

**Pull request merged**
